test/unit/HTTP_reveice_measures-test_multiple_measurement.js

```diff
@@ -0,0 +1,147 @@
+ * License along with iotagent-json.
+ * If not, seehttp://www.gnu.org/licenses/.
+ *
+ * For those usages not covered by the GNU Affero General Public License
+ * please contact with::[contacto@tid.es]
+ */
+
+/* jshint camelcase: false */
+'use strict';
+
+var iotagentMqtt = require('../../'),
+    config = require('../config-test.js'),
+    nock = require('nock'),
+    iotAgentLib = require('iotagent-node-lib'),
+    should = require('should'),
+    async = require('async'),
+    request = require('request'),
+    utils = require('../utils'),
+    groupCreation = {
+        url: 'http://localhost:' + config.iota.server.port + '/iot/services',
+        method: 'POST',
+        json: {
+            services: [
+                {
+                    resource: '/iot/json',
+                    apikey: 'KL223HHV8732SFL1',
+                    entity_type: 'TheLightType',
+                    trust: '8970A9078A803H3BL98PINEQRW8342HBAMS',
+                    cbHost: 'http://unexistentHost:1026',
+                    commands: [],
+                    lazy: [],
+                    attributes: [
+                        {
+                            name: 'status',
+                            type: 'Boolean'
+                        }
+                    ],
+                    static_attributes: []
+                }
+            ]
+        },
+        headers: {
+            'fiware-service': 'smartGondor',
+            'fiware-servicepath': '/gardens'
+        }
+    },
+    contextBrokerMock,
+    contextBrokerUnprovMock;
+
+describe('HTTP: Measure reception ', function() {
+    beforeEach(function(done) {
+        var provisionOptions = {
+            url: 'http://localhost:' + config.iota.server.port + '/iot/devices',
+            method: 'POST',
+            json: utils.readExampleFile('./test/deviceProvisioning/provisionDeviceHTTP.json'),
+            headers: {
+                'fiware-service': 'smartGondor',
+                'fiware-servicepath': '/gardens'
+            }
+        };
+
+        nock.cleanAll();
+
+        contextBrokerMock = nock('http://192.168.1.1:1026')
+            .matchHeader('fiware-service', 'smartGondor')
+            .matchHeader('fiware-servicepath', '/gardens')
+            .post('/v1/updateContext')
+            .reply(200, utils.readExampleFile('./test/contextResponses/multipleMeasuresSuccess.json'));
+
+        iotagentMqtt.start(config, function() {
+            request(provisionOptions, function(error, response, body) {
+                done();
+            });
+        });
+    });
+
+    afterEach(function(done) {
+        nock.cleanAll();
+
+        async.series([iotAgentLib.clearAll, iotagentMqtt.stop], done);
+    });
+
+    describe('When a POST measure arrives for the HTTP binding_4_multiple_measurements', function() {
+        var optionsMeasure = {
+            url: 'http://localhost:' + config.http.port + '/iot/json',
+            method: 'POST',
+            json:[ {
```

Comment:
How many requests are sent to CB when a multimeasure like this is processed? One request per measure (4 requests in this case)? Or a single request with a "sum" of all the measures?

Reply:
Multi measure means one request will have multiple measurement data. SUch fucntionality is already implemented in iotagent-ul.
Here request is 1 but we are sending more than one measure (it can be 4 measure in this case).

Reply:
Ok, one request.

Next question is, what is sent to CB when the same measure-name overlaps in different measures. In this case, humidity and temperature overlap as they appear in the four measures. I understand that the last one is the one send to CB.

In this case, the last one is humidity 35 and temperature 90 so I understand that CB will receive:

```

  "contextElements": [
    {
      "type": "AnMQTTDevice",
      "isPattern": "false",
      "id": "Second MQTT Device",
      "attributes": [
        {
          "name": "humidity",
          "type": "degrees",
          "value": "35"
        },
        {
          "name": "temperature",
          "type": "celsius",
          "value": "90"
        }
      ]
    }
  ],
  "updateAction": "UPDATE"
}
```

Is my interpretation correct?

Reply:
> Ok, one request.
> 
> Next question is, what is sent to CB when the same measure-name overlaps in different measures. In this case, humidity and temperature overlap as they appear in the four measures. I understand that the last one is the one send to CB.
> 
> In this case, the last one is humidity 35 and temperature 90 so I understand that CB will receive:
> 
> ```
> 
>   "contextElements": [
>     {
>       "type": "AnMQTTDevice",
>       "isPattern": "false",
>       "id": "Second MQTT Device",
>       "attributes": [
>         {
>           "name": "humidity",
>           "type": "degrees",
>           "value": "35"
>         },
>         {
>           "name": "temperature",
>           "type": "celsius",
>           "value": "90"
>         }
>       ]
>     }
>   ],
>   "updateAction": "UPDATE"
> }
> ```
> 
> Is my interpretation correct?

Yes your understanding correct.

Reply:
However looking to the mocks expectations at 

```
.post('/v1/updateContext', utils.readExampleFile('./test/contextRequests/multipleMeasures.json'))
```

The [multipleMeasures.json](https://github.com/telefonicaid/iotagent-json/blob/master/test/contextRequests/multipleMeasures.json) is:

```
{
  "contextElements": [
    {
      "type": "AnMQTTDevice",
      "isPattern": "false",
      "id": "Second MQTT Device",
      "attributes": [
        {
          "name": "humidity",
          "type": "degrees",
          "value": "32"
        },
        {
          "name": "temperature",
          "type": "celsius",
          "value": "87"
        }
      ]
    }
  ],
  "updateAction": "UPDATE"
}
```

That is humidity=32 and temperature=87 and not humidity=35 and temperature=90 as it should be.

Thus, I understand something is not working as expected... What do you think?

Reply:
> However looking to the mocks expectations at
> 
> ```
> .post('/v1/updateContext', utils.readExampleFile('./test/contextRequests/multipleMeasures.json'))
> ```
> 
> The [multipleMeasures.json](https://github.com/telefonicaid/iotagent-json/blob/master/test/contextRequests/multipleMeasures.json) is:
> 
> ```
> {
>   "contextElements": [
>     {
>       "type": "AnMQTTDevice",
>       "isPattern": "false",
>       "id": "Second MQTT Device",
>       "attributes": [
>         {
>           "name": "humidity",
>           "type": "degrees",
>           "value": "32"
>         },
>         {
>           "name": "temperature",
>           "type": "celsius",
>           "value": "87"
>         }
>       ]
>     }
>   ],
>   "updateAction": "UPDATE"
> }
> ```
> 
> That is humidity=32 and temperature=87 and not humidity=35 and temperature=90 as it should be.
> 
> Thus, I understand something is not working as expected... What do you think?

The last measurement is the my expected value to update in CB. however i've again verified IotAgent Json with below data:

**Sensor data passed to IOT:**
[
{
"temperature": "50",
"humidity": "30"
},
{
"temperature": "60",
"humidity": "40"
}
]

**Measurment shown on CB:**
[
    {
        "id": "multiMeasure",
        "type": "multiType",
        "TimeInstant": {
            "type": "ISO8601",
            "value": "2019-06-20T07:44:23.00Z",
            "metadata": {}
        },
        "humidity": {
            "type": "string",
            "value": "40",
            "metadata": {
                "TimeInstant": {
                    "type": "ISO8601",
                    "value": "2019-06-20T07:44:23.750Z"
                }
            }
        },
        "temperature": {
            "type": "string",
            "value": "60",
            "metadata": {
                "TimeInstant": {
                    "type": "ISO8601",
                    "value": "2019-06-20T07:44:23.750Z"
                }
            }
        }
    }
]

Reply:
Maybe we are not understanding each other :) I'll try again, from the beginning.

This unit test case models a situation in which you send a multi-measures with 4 items to the IOTA and check the request that CB will receive in that case. Not with a real CB but with a mock to check request expectations (I understand you are familiar with the general concepts of unit testing and mocks, otherwise please tell me).

As far as I understand by the previous conversation (maybe I have interpreted something incorrectly, in that case, please tell me), when the IOTA receives a multi-measure with N elements all them are "summed together" so only 1 request is sent to the CB. Moreover, if the same attribute appears in several measures (as humidity and temperature in this case) the one corresponding to the last measure is the one which overrides the others. In the particular example of this test the last measure is humidity: '35' and temperature: '90'.

Thus CB should receive a request with humidity: '35' and temperature: '90', shouldn't it?

**However**, let's analyze the expectations that are being used in the mock. That's the line:

```
.post('/v1/updateContext', utils.readExampleFile('./test/contextRequests/multipleMeasures.json'))
```

and if we look to that test/contextRequests/multipleMeasures.json file in the repository we have:

```
{
  "contextElements": [
    {
      "type": "AnMQTTDevice",
      "isPattern": "false",
      "id": "Second MQTT Device",
      "attributes": [
        {
          "name": "humidity",
          "type": "degrees",
          "value": "32"
        },
        {
          "name": "temperature",
          "type": "celsius",
          "value": "87"
        }
      ]
    }
  ],
  "updateAction": "UPDATE"
}
```

That is temperature 87 and humidity 32. And 35 != 32 and 90 != 87. Thus, in summary, the test is not correct (in fact, I wonder how travis is passing this as ok...)

I hope to have clarified the problem... Please confirm you understand it (or tell what is wrong in my discussion, maybe I have missed something :).

Reply:
> Maybe we are not understanding each other :) I'll try again, from the beginning.
> 
> This unit test case models a situation in which you send a multi-measures with 4 items to the IOTA and check the request that CB will receive in that case. Not with a real CB but with a mock to check request expectations (I understand you are familiar with the general concepts of unit testing and mocks, otherwise please tell me).
> 
> As far as I understand by the previous conversation (maybe I have interpreted something incorrectly, in that case, please tell me), when the IOTA receives a multi-measure with N elements all them are "summed together" so only 1 request is sent to the CB. Moreover, if the same attribute appears in several measures (as humidity and temperature in this case) the one corresponding to the last measure is the one which overrides the others. In the particular example of this test the last measure is humidity: '35' and temperature: '90'.
> 
> Thus CB should receive a request with humidity: '35' and temperature: '90', shouldn't it?
> 
> **However**, let's analyze the expectations that are being used in the mock. That's the line:
> 
> ```
> .post('/v1/updateContext', utils.readExampleFile('./test/contextRequests/multipleMeasures.json'))
> ```
> 
> and if we look to that test/contextRequests/multipleMeasures.json file in the repository we have:
> 
> ```
> {
>   "contextElements": [
>     {
>       "type": "AnMQTTDevice",
>       "isPattern": "false",
>       "id": "Second MQTT Device",
>       "attributes": [
>         {
>           "name": "humidity",
>           "type": "degrees",
>           "value": "32"
>         },
>         {
>           "name": "temperature",
>           "type": "celsius",
>           "value": "87"
>         }
>       ]
>     }
>   ],
>   "updateAction": "UPDATE"
> }
> ```
> 
> That is temperature 87 and humidity 32. And 35 != 32 and 90 != 87. Thus, in summary, the test is not correct (in fact, I wonder how travis is passing this as ok...)
> 
> I hope to have clarified the problem... Please confirm you understand it (or tell what is wrong in my discussion, maybe I have missed something :).

Thanks for the explanation.
I understood your point.
The mock object contextBrokerMock of beforeEach block written in "HTTP_reveice_measures-test_multiple_measurement.js" is not used in the execution of below test case
        it('should return a 200 OK with no error', function() {
            request(optionsMeasure, function(error, result, body) {
                should.not.exist(error);
                result.statusCode.should.equal(200);
                //done();
            });
        });
So the expectation of multipleMeasure.json is not as per request data provided in optionsMeasure.
In the UT it test only returning of no error and sttaus code as 200.
From UT  code, beforeEach block should be removed. I'll update this.
Please let me know if more explanation is required.

Reply:
I didn't realize about the commented line `//done()` :O. Please, don't do that. The testing framework needs done() to be called after test termination. Not doing so can bring unexpected problems. This documentation may help to get the point: https://mochajs.org/#asynchronous-code

> The mock object contextBrokerMock of beforeEach block written in "HTTP_reveice_measures-test_multiple_measurement.js" is not used in the execution of below test case

This shoulf be fixed. The test has to check the expectaions of the mock who plays the role of CB. Otherwise, the test is pointless :)

In particular what I mean is:

1. Create a JSON file similar to multipleMeasures.json but modeling your expected case (humidity 35 and temperature 90). Name it for instance multipleMeasures2.json
2. Use that file in contextBrokerMock() and get the test passed
3. Check that if you change something in multipleMeasures2.json (e.g. humidity set to 30) your test fails